go/vt/vttablet/tabletserver/throttle/throttler.go

```diff
@@ -0,0 +1,687 @@
+			return
+		}
+		log.Infof("Throttler: read instance key: %+v", key)
+
+		probe := &mysql.Probe{
+			Key:         *key,
+			User:        clusterSettings.User,
+			Password:    clusterSettings.Password,
+			MetricQuery: clusterSettings.MetricQuery,
+			CacheMillis: clusterSettings.CacheMillis,
+		}
+		(*probes)[*key] = probe
+	}
+
+	for clusterName, clusterSettings := range config.Settings().Stores.MySQL.Clusters {
+		clusterName := clusterName
+		clusterSettings := clusterSettings
+		// config may dynamically change, but internal structure (config.Settings().Stores.MySQL.Clusters in our case)
+		// is immutable and can only be _replaced_. Hence, it's safe to read in a goroutine:
+		go func() error {
+			throttler.mysqlClusterThresholds.Set(clusterName, clusterSettings.ThrottleThreshold, cache.DefaultExpiration)
+
+			tabletAliases, err := throttler.ts.FindAllTabletAliasesInShard(ctx, throttler.keyspace, throttler.shard)
+			if err != nil {
+				return err
+			}
+			clusterProbes := &mysql.ClusterProbes{
+				ClusterName:      clusterName,
+				IgnoreHostsCount: clusterSettings.IgnoreHostsCount,
+				InstanceProbes:   mysql.NewProbes(),
+			}
+			for _, tabletAlias := range tabletAliases {
+				tablet, err := throttler.ts.GetTablet(ctx, tabletAlias)
+				if err != nil {
+					return err
+				}
+				if tablet.Type == topodatapb.TabletType_REPLICA {
+					key := mysql.InstanceKey{Hostname: tablet.MysqlHostname, Port: int(tablet.MysqlPort)}
+					addInstanceKey(&key, clusterName, clusterSettings, clusterProbes.InstanceProbes)
+				}
+			}
+			throttler.mysqlClusterProbesChan <- clusterProbes
+			return nil
+		}()
+	}
+	return nil
+}
+
+// synchronous update of inventory
+func (throttler *Throttler) updateMySQLClusterProbes(ctx context.Context, clusterProbes *mysql.ClusterProbes) error {
+	log.Infof("Throttler: updating MySQLClusterProbes: %s", clusterProbes.ClusterName)
+	throttler.mysqlInventory.ClustersProbes[clusterProbes.ClusterName] = clusterProbes.InstanceProbes
+	throttler.mysqlInventory.IgnoreHostsCount[clusterProbes.ClusterName] = clusterProbes.IgnoreHostsCount
+	throttler.mysqlInventory.IgnoreHostsThreshold[clusterProbes.ClusterName] = clusterProbes.IgnoreHostsThreshold
+	return nil
+}
+
+// synchronous aggregation of collected data
+func (throttler *Throttler) aggregateMySQLMetrics(ctx context.Context) error {
+	for clusterName, probes := range throttler.mysqlInventory.ClustersProbes {
+		metricName := fmt.Sprintf("mysql/%s", clusterName)
+		ignoreHostsCount := throttler.mysqlInventory.IgnoreHostsCount[clusterName]
+		ignoreHostsThreshold := throttler.mysqlInventory.IgnoreHostsThreshold[clusterName]
+		aggregatedMetric := aggregateMySQLProbes(ctx, probes, clusterName, throttler.mysqlInventory.InstanceKeyMetrics, ignoreHostsCount, config.Settings().Stores.MySQL.IgnoreDialTCPErrors, ignoreHostsThreshold)
+		go throttler.aggregatedMetrics.Set(metricName, aggregatedMetric, cache.DefaultExpiration)
```

Comment:
is this intentional? is there benefit of doing it in go routine?

Reply:
It's a bit meaningless since the cache is thread-safe. I'll remove the go routine.

Reply:
removed.